koschei/frontend/api.py

```diff
@@ -0,0 +1,52 @@
+# Copyright (C) 2017 Red Hat, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Author: Mikolaj Izdebski <mizdebsk@redhat.com>
+
+from flask import jsonify, request
+from sqlalchemy.orm import joinedload
+from koschei.frontend import app, db
+from koschei.models import Package, Collection
+
+
+def build_to_json(build):
+    if build:
+        return {'task_id': build.task_id}
+
+
+def package_to_json(package):
+    if package:
+        return {'name': package.name,
+                  'collection': package.collection.name,
+                  'state': package.state_string,
+                  'last_complete_build': build_to_json(package.last_complete_build)}
+
+
+def packages_to_json(packages):
+    return [package_to_json(p) for p in packages]
+
+
+@app.route('/api/v1/packages')
+def list_packages():
+    query = db.query(Package).join(Collection)\
+        .options(joinedload(Package.collection))\
+        .options(joinedload(Package.last_complete_build))
+    if 'name' in request.args:
+        query = query.filter(Package.name.in_(request.args.getlist('name')))
+    if 'collection' in request.args:
+        query = query.filter(Collection.name.in_(request.args.getlist('collection')))
+    packages = query.order_by(Package.name).all()
```

Comment:
The query is still not right. It works, but it takes half a second to execute.
- `join` and `joinedload` are two separate things that don't cooperate. So you end up with two joins. When you need the functionality of both at the same time, you need to use `contains_eager` instead of `joinedload`.
- `join` emits a left outer join, for joins on non-nullable fields it's better to use `innerjoin`
- In this case, I wouldn't use any of those. Frontend already has all the collections in memory, bound to `g.collections`. There's also `g.collections_by_name` dict, so I would use that to get the ids and filter by `Package.collection_id`.